### PR TITLE
chore: Explicitly list packages for bundling

### PR DESCRIPTION
**Pull request by dmaretskyi**
### Summary
📦🛠️🌐

This pull request updates and adds some dependencies for various DXOS packages, and improves the browser bundling of some packages that use node modules. It also uses the `node:` prefix for importing built-in node modules in the `plate` and `log` packages. Additionally, it adds a dependency on `@dxos/react-surface` to two plugins that use the UI component library.

> _We're building the browser bundle of doom_
> _Stripping out node modules with `empty.ts`_
> _Adding UI components and echo schemas_
> _Bundling `hypercore` and `node-std`_

### Walkthrough
*  Add `@dxos/react-surface` as a dependency to `plugin-markdown` and `plugin-stack` packages to use common UI components ([link](https://github.com/dxos/dxos/pull/3530/files?diff=unified&w=0#diff-e6d1d345821614148e5995e27dd43e3d23acd9c585a2ae3ef5911c496baf2ccdR36), [link](https://github.com/dxos/dxos/pull/3530/files?diff=unified&w=0#diff-b0988acb4500e2e188ebdfee4ad90811c200cd24700f45e3534f33ebe048cdf3R36))
* Add `@faker-js/faker` as a dependency to `feed-store` package to generate fake data for testing and mocking ([link](https://github.com/dxos/dxos/pull/3530/files?diff=unified&w=0#diff-b61df366c52f452d49c6a479e47aef3c33a642e65c87da080208a5d330f79320R34))
* Add and remove `@dxos/keyring` as a dependency to `feed-store` package to manage cryptographic keys and identities ([link](https://github.com/dxos/dxos/pull/3530/files?diff=unified&w=0#diff-b61df366c52f452d49c6a479e47aef3c33a642e65c87da080208a5d330f79320R41), [link](https://github.com/dxos/dxos/pull/3530/files?diff=unified&w=0#diff-b61df366c52f452d49c6a479e47aef3c33a642e65c87da080208a5d330f79320L45-R47))
* Modify build options for `hypercore` and `node-std` packages to reduce browser bundle size and complexity by aliasing and bundling some node modules ([link](https://github.com/dxos/dxos/pull/3530/files?diff=unified&w=0#diff-494b5cee5869acf00c637408cc22abee3e9420f87f767b94c12061663166d803L19-R73), [link](https://github.com/dxos/dxos/pull/3530/files?diff=unified&w=0#diff-23465bddeb6db0d95f2be21dbda2d2753bfcb195c7f8dbf9c98b06e1a1b718d9L10-R41))
* Add `empty.ts` file to `hypercore` package to stub out some packages that are stripped from the browser bundle ([link](https://github.com/dxos/dxos/pull/3530/files?diff=unified&w=0#diff-37e4ceffb13c29017785c0e944847e152a86b6df88569b4f0696089bfadf55fcL1-R6))
* Update imports of node modules in `log` and `plate` packages to use `node:` prefix to avoid confusion and conflicts with external modules ([link](https://github.com/dxos/dxos/pull/3530/files?diff=unified&w=0#diff-62ea19dc18973d13ed4022fffa35f2409d38054616157e0389965a9dcf0643a2L5-R7), [link](https://github.com/dxos/dxos/pull/3530/files?diff=unified&w=0#diff-d02472efac05b4664724113ba7a4e63014690f88efe6c56469c341bb91f3cbe9L6-R6), [link](https://github.com/dxos/dxos/pull/3530/files?diff=unified&w=0#diff-67f1a701dd88452f6fcedbf6733e4cc216311c5e41acbd0a6eee5357aa74df46L6-R7), [link](https://github.com/dxos/dxos/pull/3530/files?diff=unified&w=0#diff-405faff1f2f86c1137f755cd1f7e7aaf103e5adc60666561475a18dd8840b668L4-R4), [link](https://github.com/dxos/dxos/pull/3530/files?diff=unified&w=0#diff-29554b73a99dc7b8db794f69cdc8b987236c5a5cee9a39d103f352c2ce7139b1L5-R7), [link](https://github.com/dxos/dxos/pull/3530/files?diff=unified&w=0#diff-5c9b293bf89c99763582ba79b42078a9562951e4704806abed2903d17c24ce30L5-R6), [link](https://github.com/dxos/dxos/pull/3530/files?diff=unified&w=0#diff-0b3154fc25dbf622215a85ec70369da851bc4c6489aeb53e285d786a902ec5d2L5-R6), [link](https://github.com/dxos/dxos/pull/3530/files?diff=unified&w=0#diff-f04b15e4a289b08166a53091f246425c155e972da084aebf8f3ed66a9a262240L7-R7), [link](https://github.com/dxos/dxos/pull/3530/files?diff=unified&w=0#diff-2a3a55b805e9c003d73296eef966e512a20ccc893ee83bea58171b0a741187a0L5-R7), [link](https://github.com/dxos/dxos/pull/3530/files?diff=unified&w=0#diff-266e1f39d37556f2258013c1cfc9ba97370dadc3e90c2799d81b7881b32ab20bL5-R6), [link](https://github.com/dxos/dxos/pull/3530/files?diff=unified&w=0#diff-ff97a55e96fb3c6240c74ddbbbba298c1e6ef6dce5f6f424c2ec0a10d164ac56L7-R7))
* Add `react` as a peer dependency to `observable-object` package to avoid bundling multiple versions of `react` or causing version conflicts ([link](https://github.com/dxos/dxos/pull/3530/files?diff=unified&w=0#diff-fd5dbe48a14c499a3936b9d4d99968aaa557b544ee4d4acb431637d1ec6b97d7R51-R53))
* Add `@dxos/echo-pipeline` as a dependency to `echo-schema` package to provide utilities for working with echo schemas and types ([link](https://github.com/dxos/dxos/pull/3530/files?diff=unified&w=0#diff-d4cf342a7b74c84eea38b44b64c7ecc318e86e260aee4e9e93ef34e75db3167cR32))
* Add `@dxos/protobuf-compiler` as a dependency to `echo-typegen` package to provide a custom compiler for echo protobuf definitions ([link](https://github.com/dxos/dxos/pull/3530/files?diff=unified&w=0#diff-27fe718b163c6743d69743c7413bc196ab8bd1350bef48107f9890605ecf45f4R23))

### CODE REVIEW

File: tools/executors/esbuild/src/main.ts [hunk 1]
```diff
@@ -35,12 +36,14 @@ export default async (options: EsbuildExecutorOptions, context: ExecutorContext)
...
     await readdir(options.outputPath);
     await rm(options.outputPath, { recursive: true });
-  } catch {}
+  } catch { }
 
   // TODO(wittjosiah): Workspace from context is deprecated.
   const packagePath = join(context.workspace!.projects[context.projectName!].root, 'package.json');
   const packageJson = JSON.parse(await readFile(packagePath, 'utf-8'));
 
+  const runtimeDeps = new Set([...Object.keys(packageJson.dependencies ?? {}), ...Object.keys(packageJson.peerDependencies ?? {})]);
```
Comment (wittjosiah) [inline tools/executors/esbuild/src/main.ts:45]: why do this outside the plugin? feels like something the plugin should know how to do given the resolveDir

File: tools/executors/esbuild/src/bundle-deps-plugin.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+  /**
+   * Runtime dependencies available to the bundle.
+   */
+  runtimeDeps: Set<string>;
+
+  resolveDir: string;
+}
+
+export const bundleDepsPlugin = (options: BundleDepsPluginOptions): Plugin => ({
```
Comment (wittjosiah) [inline tools/executors/esbuild/src/bundle-deps-plugin.ts:22]:
```suggestion
/**
 * Ensures all external dependencies are marked as external unless specifically listed for being included in the package bundle. 
 */
export const bundleDepsPlugin = (options: BundleDepsPluginOptions): Plugin => ({
```

File: packages/sdk/react-client/package.json [hunk 3]
```diff
@@ -29,15 +30,15 @@
     "@dxos/react-async": "workspace:*",
     "@dxos/rpc": "workspace:*",
     "@dxos/rpc-tunnel": "workspace:*",
+    "@dxos/text-model": "workspace:*",
     "@dxos/util": "workspace:*",
+    "@faker-js/faker": "^7.3.0",
```
Comment (wittjosiah) [inline packages/sdk/react-client/package.json:35]: hmmm concerned about this, I'm assuming you moved all deps from dev deps that the new plugin found were required at run time? Perhaps some of these should be peer or even optional dependencies

File: packages/common/node-std/project.json [hunk 4]
```diff
@@ -7,12 +7,38 @@
       "options": {
         "bundlePackages": [
           "assert",
+          "available-typed-arrays",
```
Comment (wittjosiah) [inline packages/common/node-std/project.json:10]: wow, why are all these needed? are transitive deps not automatically bundled? I suppose that makes sense...

File: packages/common/hypercore/src/empty.ts [hunk 5]
```diff
@@ -0,0 +1,13 @@
+// Empty module to stub stripped packages.
```
Comment (wittjosiah) [inline packages/common/hypercore/src/empty.ts:1]: what is a "stripped package"?

File: packages/sdk/react-client/package.json (hunk 3 above)
Comment (dmaretskyi) [reply packages/sdk/react-client/package.json:35]: yes, I think a lot of the ones that were moved are only used in /testing entrypoint. The downside of moving it here is dependency bloat at install time, but they should not be included in a built bundle.

We can do a pass to clean this up and move them to peer/optional deps if necessary

**Pull request merged**